src/bitprim/keoken/memory_state.cpp

```diff
@@ -29,15 +29,82 @@ using libbitcoin::wallet::payment_address;
 namespace bitprim {
 namespace keoken {
 
-// state::state(asset_id_t asset_id_initial)
+// memory_state::memory_state(asset_id_t asset_id_initial)
 //     : asset_id_next_(asset_id_initial)
 // {}
 
-void state::set_initial_asset_id(asset_id_t asset_id_initial) {
+void memory_state::set_initial_asset_id(asset_id_t asset_id_initial) {
+    asset_id_initial_ = asset_id_initial;
     asset_id_next_ = asset_id_initial;
 }
 
-void state::create_asset(std::string asset_name, amount_t asset_amount, 
+// private
+template <typename Predicate>
+void memory_state::remove_balance_entries(Predicate const& pred) {
+    // precondition: mutex must be locked
+
+    // for(auto&& x : balance_) {
+    //     auto& balance_entries = x.second;
+
+    //     balance_entries.erase(
+    //         std::remove_if(begin(balance_entries), end(balance_entries), pred), 
+    //     end(balance_entries));
+    // }
+
+    auto f = begin(balance_); 
+    auto l = end(balance_);
+
+    while (f != l) {
+        auto& balance_entries = f->second;
+
+        balance_entries.erase(
+            std::remove_if(begin(balance_entries), end(balance_entries), pred), 
+        end(balance_entries));
+
+        if (balance_entries.empty()) {
+            f = balance_.erase(f);
+        } else {
+            ++f;
+        }
+    }
+}
+
+void memory_state::reset() {
+    boost::unique_lock<boost::shared_mutex> lock(mutex_);
+
+    asset_id_next_ = asset_id_initial_;
+    asset_list_.clear();
+    balance_.clear();
+}
+
+struct rollback_pred {
+    size_t height;
+
+    explicit
+    rollback_pred(size_t height)
+        : height(height)
+    {}
+
+    template <typename Entry>
+    bool operator()(Entry const& entry) const {
+        return entry.block_height > height;
+    }
+};
+
+void memory_state::rollback_to(size_t height) {
+    boost::unique_lock<boost::shared_mutex> lock(mutex_);
+
+    remove_balance_entries(rollback_pred{height});
+
+    asset_list_.erase(
+        std::remove_if(begin(asset_list_), end(asset_list_), rollback_pred{height}), 
+    end(asset_list_));
+
+    asset_id_next_ = asset_list_.empty() ? asset_id_initial_ : asset_list_.back().asset.id();
```

Comment:
Missing asset_list_.back().asset.id() + 1 ?

Reply:
Yeah, you are right! 
Good catch! 
Thanks!
`asset_id_next_ = asset_list_.empty() ? asset_id_initial_ : asset_list_.back().asset.id() + 1;`